src/osd/OSD.cc

```diff
@@ -9909,11 +9972,9 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
   osd->service.maybe_inject_dispatch_delay();
 
   // [lookup +] lock pg (if we have it)
-  if (!pg) {
-    pg = osd->_lookup_lock_pg(item.first);
-  } else {
-    pg->lock();
-  }
+  PGQueueable::OrderLocker::Ref l;
+  l = item.second.get_order_locker(pg, item.first, osd);
+  l->lock();
 
```

Comment:
@myoungwon i just finished rebasing Sam's wip-osd-reply at https://github.com/tchaikov/ceph/tree/wip-osd-reply. the commit of https://github.com/tchaikov/ceph/commit/edf07b9708f349931026eb9d62a158a603e39f7e compiles, but the latest one (https://github.com/tchaikov/ceph/commit/f98a71a241d97ec9647469078f30c878bb0de1d3) is still a WIP. if you want to continue working on this, please feel free to base your work on it. and i will leave it to you.